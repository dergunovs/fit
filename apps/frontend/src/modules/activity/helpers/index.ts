import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
