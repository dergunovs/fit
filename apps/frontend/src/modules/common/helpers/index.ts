import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
