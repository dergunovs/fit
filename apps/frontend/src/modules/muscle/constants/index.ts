export const URL_MUSCLE = '/muscles';
export const URL_MUSCLE_CREATE = `${URL_MUSCLE}/create`;
export const URL_MUSCLE_EDIT = `${URL_MUSCLE}/edit`;

export const MUSCLE_LIST_HEADERS = [{ title: 'Мышцы' }, { title: 'Цвет' }];
export const MUSCLE_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
