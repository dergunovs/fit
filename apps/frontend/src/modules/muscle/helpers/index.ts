import { IExerciseDone, IMuscle, IMuscleStatistics } from 'fitness-tracker-contracts';

export function generateMuscleStatistics(exercises: IExerciseDone[], muscles: IMuscle[]) {
  const muscleStatistics: IMuscleStatistics[] = [];

  muscles.forEach((muscle: IMuscle) => {
    const title = muscle.title;
    const color = muscle.color || '#000';
    let sets = 0;
    let repeats = 0;

    exercises.forEach((exercise: IExerciseDone) => {
      const setsCount =
        exercise.exercise?.muscles?.filter((muscleToFilter) => muscleToFilter._id === muscle._id).length || 0;

      sets += setsCount;

      if (exercise.exercise?.muscles?.some((groupToFilter) => groupToFilter._id === muscle._id)) {
        repeats += exercise.repeats;
      }
    });

    if (sets) muscleStatistics.push({ title, color, sets, repeats });
  });

  return muscleStatistics.sort((a, b) => b.repeats - a.repeats);
}
