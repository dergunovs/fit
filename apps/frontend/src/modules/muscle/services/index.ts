import { ComputedRef } from 'vue';
import {
  API_MUSCLE,
  TGetMusclesDTO,
  TGetMuscleDTO,
  TPostMuscleDTO,
  TPostMuscleDataDTO,
  TUpdateMuscleDTO,
  TUpdateMuscleDataDTO,
  TDeleteMuscleDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const muscleService = {
  getAll: () => {
    return useQuery({
      queryKey: [API_MUSCLE],
      queryFn: async () => {
        const { data } = await api.get<TGetMusclesDTO>(API_MUSCLE);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_MUSCLE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetMuscleDTO>(`${API_MUSCLE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_MUSCLE],
      mutationFn: async (formData: TPostMuscleDataDTO) => {
        const { data } = await api.post<TPostMuscleDTO>(API_MUSCLE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_MUSCLE],
      mutationFn: async (formData: TUpdateMuscleDataDTO) => {
        const { data } = await api.patch<TUpdateMuscleDTO>(`${API_MUSCLE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_MUSCLE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteMuscleDTO>(`${API_MUSCLE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
