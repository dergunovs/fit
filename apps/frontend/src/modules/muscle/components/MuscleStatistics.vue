<template>
  <div :class="$style.table">
    <UiTable :headers="MUSCLE_GROUP_STATISTICS_HEADERS">
      <tr v-for="muscle in props.statistics" :key="muscle.title" data-test="muscle-row">
        <td :style="{ color: muscle.color }" data-test="muscle-title">{{ muscle.title }}</td>
        <td data-test="muscle-sets">{{ muscle.sets }}</td>
        <td data-test="muscle-repeats">{{ muscle.repeats }}</td>
      </tr>
    </UiTable>
  </div>
</template>

<script setup lang="ts">
import { IMuscleStatistics } from 'fitness-tracker-contracts';
import { UiTable } from 'mhz-ui';

import { MUSCLE_GROUP_STATISTICS_HEADERS } from '@/exercise/constants';

interface IProps {
  statistics: IMuscleStatistics[];
}

const props = defineProps<IProps>();
</script>

<style module lang="scss">
.table {
  display: table;
}
</style>
