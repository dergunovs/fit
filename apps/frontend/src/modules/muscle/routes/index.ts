import { URL_MUSCLE, URL_MUSCLE_CREATE, URL_MUSCLE_EDIT } from '@/muscle/constants';

export const muscleRoutes = [
  {
    path: URL_MUSCLE,
    name: 'MuscleList',
    component: () => import('@/muscle/pages/MuscleListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: URL_MUSCLE_CREATE,
    name: 'MuscleCreate',
    component: () => import('@/muscle/pages/MuscleCreatePage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_MUSCLE_EDIT}/:muscle`,
    name: 'MuscleEdit',
    component: () => import('@/muscle/pages/MuscleEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
