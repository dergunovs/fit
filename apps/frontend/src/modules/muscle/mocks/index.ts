import { vi } from 'vitest';
import {
  TDeleteMuscleDTO,
  TPostMuscleDataDTO,
  TPostMuscleDTO,
  TUpdateMuscleDataDTO,
  TUpdateMuscleDTO,
} from 'fitness-tracker-contracts';

import { muscleService } from '@/muscle/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { MUSCLE_FIXTURE, MUSCLES_FIXTURE } from '@/muscle/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetMuscle = vi.spyOn(muscleService, 'getOne').mockReturnValue(mockQueryReply(MUSCLE_FIXTURE));
const spyGetMuscles = vi.spyOn(muscleService, 'getAll').mockImplementation(() => mockQueryReply(MUSCLES_FIXTURE));

const spyCreateMuscle = vi.fn();
const spyUpdateMuscle = vi.fn();
const spyDeleteMuscle = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(muscleService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostMuscleDTO, TPostMuscleDataDTO>(spyCreateMuscle);
});

vi.spyOn(muscleService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateMuscleDTO, TUpdateMuscleDataDTO>(spyUpdateMuscle);
});

vi.spyOn(muscleService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteMuscleDTO, string>(spyDeleteMuscle);
});

export { spyGetMuscle, spyGetMuscles, spyCreateMuscle, spyUpdateMuscle, spyDeleteMuscle, mockOnSuccess };
