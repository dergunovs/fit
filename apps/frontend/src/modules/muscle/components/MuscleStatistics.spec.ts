import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import MuscleStatistics from './MuscleStatistics.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISES_DONE_FIXTURE } from '@/exercise/fixtures';
import { generateMuscleStatistics } from '@/muscle/helpers';
import { MUSCLES_FIXTURE } from '@/muscle/fixtures';

const tableRow = dataTest('muscle-row');
const title = dataTest('muscle-title');
const sets = dataTest('muscle-sets');
const repeats = dataTest('muscle-repeats');

let wrapper: VueWrapper<InstanceType<typeof MuscleStatistics>>;

const statistics = generateMuscleStatistics(EXERCISES_DONE_FIXTURE, MUSCLES_FIXTURE);

beforeEach(() => {
  wrapper = wrapperFactory(MuscleStatistics, { statistics });
});

enableAutoUnmount(afterEach);

describe('MuscleStatistics', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(MuscleStatistics)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows exercises in table', async () => {
    expect(wrapper.findAll(tableRow).length).toBe(statistics.length);

    expect(wrapper.find(title).text()).toBe(statistics[0].title);
    expect(wrapper.find(sets).text()).toBe(statistics[0].sets.toString());
    expect(wrapper.find(repeats).text()).toBe(statistics[0].repeats.toString());
  });
});
