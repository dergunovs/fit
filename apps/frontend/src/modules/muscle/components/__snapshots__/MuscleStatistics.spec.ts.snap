// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MuscleStatistics > matches snapshot 1`] = `
"<div class="_table_b81ac8">
  <table headers="[object Object],[object Object],[object Object]" lang="ru">
    <tbody>
      <tr data-test="muscle-row">
        <td style="color: #00CED1;" data-test="muscle-title">Грудь</td>
        <td data-test="muscle-sets">1</td>
        <td data-test="muscle-repeats">12</td>
      </tr>
      <tr data-test="muscle-row">
        <td style="color: #000080;" data-test="muscle-title">Ноги</td>
        <td data-test="muscle-sets">1</td>
        <td data-test="muscle-repeats">12</td>
      </tr>
      <tr data-test="muscle-row">
        <td style="color: #C41E3A;" data-test="muscle-title">Плечи</td>
        <td data-test="muscle-sets">1</td>
        <td data-test="muscle-repeats">12</td>
      </tr>
      <tr data-test="muscle-row">
        <td style="color: #32CD32;" data-test="muscle-title">Пресс</td>
        <td data-test="muscle-sets">1</td>
        <td data-test="muscle-repeats">12</td>
      </tr>
    </tbody>
  </table>
</div>"
`;
